File: ios/Sources/ACCore/Models/Elements/ChartElements.swift
import Foundation
// MARK: - Charts

public struct ChartDataPoint: Codable, Equatable, Identifiable {
    public let id: String
    public var label: String
    public var value: Double
    public var color: String?

    // Use label as stable ID (with value to make it unique if needed)
    public var id: String {
        "\(label)_\(value)"
    }

    
    public init(
        label: String,
        value: Double,
        color: String? = nil,
        id: String = UUID().uuidString
    ) {
        self.id = id
        self.label = label
        self.value = value
        self.color = color
    }
    
    private enum CodingKeys: String, CodingKey {
        case label
        case value
        case color
    }
}

public struct DonutChart: Codable, Equatable {
    public let type: String = "DonutChart"
    public var id: String?
    public var title: String?
    public var data: [ChartDataPoint]
    public var colors: [String]?
    public var size: String?
    public var showLegend: Bool?
    public var innerRadiusRatio: Double?
    public var isVisible: Bool?
    public var spacing: Spacing?
    public var separator: Bool?
    public var height: BlockElementHeight?
    public var requires: [String: String]?

    public init(
        id: String? = nil,
        title: String? = nil,
        data: [ChartDataPoint],
        colors: [String]? = nil,
        size: String? = nil,
        showLegend: Bool? = nil,
        innerRadiusRatio: Double? = nil,
        isVisible: Bool? = nil,
        spacing: Spacing? = nil,
        separator: Bool? = nil,
        height: BlockElementHeight? = nil,
        requires: [String: String]? = nil
    ) {
        self.id = id
        self.title = title
        self.data = data
        self.colors = colors
        self.size = size
        self.showLegend = showLegend
        self.innerRadiusRatio = innerRadiusRatio
        self.isVisible = isVisible
        self.spacing = spacing
        self.separator = separator
        self.height = height
        self.requires = requires
    }
}

public struct BarChart: Codable, Equatable {
    public let type: String = "BarChart"
    public var id: String?
    public var title: String?
    public var data: [ChartDataPoint]
    public var colors: [String]?
    public var size: String?
    public var showLegend: Bool?
    public var orientation: String?
    public var showValues: Bool?
    public var isVisible: Bool?
    public var spacing: Spacing?
    public var separator: Bool?
    public var height: BlockElementHeight?
    public var requires: [String: String]?

    public init(
        id: String? = nil,
        title: String? = nil,
        data: [ChartDataPoint],
        colors: [String]? = nil,
        size: String? = nil,
        showLegend: Bool? = nil,
        orientation: String? = nil,
        showValues: Bool? = nil,
        isVisible: Bool? = nil,
        spacing: Spacing? = nil,
        separator: Bool? = nil,
        height: BlockElementHeight? = nil,
        requires: [String: String]? = nil
    ) {
        self.id = id
        self.title = title
        self.data = data
        self.colors = colors
        self.size = size
        self.showLegend = showLegend
        self.orientation = orientation
        self.showValues = showValues
        self.isVisible = isVisible
        self.spacing = spacing
        self.separator = separator
        self.height = height
        self.requires = requires
    }
}

public struct LineChart: Codable, Equatable {
    public let type: String = "LineChart"
    public var id: String?
    public var title: String?
    public var data: [ChartDataPoint]
    public var colors: [String]?
    public var size: String?
    public var showLegend: Bool?
    public var showDataPoints: Bool?
    public var smooth: Bool?
    public var isVisible: Bool?
    public var spacing: Spacing?
    public var separator: Bool?
    public var height: BlockElementHeight?
    public var requires: [String: String]?

    public init(
        id: String? = nil,
        title: String? = nil,
        data: [ChartDataPoint],
        colors: [String]? = nil,
        size: String? = nil,
        showLegend: Bool? = nil,
        showDataPoints: Bool? = nil,
        smooth: Bool? = nil,
        isVisible: Bool? = nil,
        spacing: Spacing? = nil,
        separator: Bool? = nil,
        height: BlockElementHeight? = nil,
        requires: [String: String]? = nil
    ) {
        self.id = id
        self.title = title
        self.data = data
        self.colors = colors
        self.size = size
        self.showLegend = showLegend
        self.showDataPoints = showDataPoints
        self.smooth = smooth
        self.isVisible = isVisible
        self.spacing = spacing
        self.separator = separator
        self.height = height
        self.requires = requires
    }
}

public struct PieChart: Codable, Equatable {
    public let type: String = "PieChart"
    public var id: String?
    public var title: String?
    public var data: [ChartDataPoint]
    public var colors: [String]?
    public var size: String?
    public var showLegend: Bool?
    public var showPercentages: Bool?
    public var isVisible: Bool?
    public var spacing: Spacing?
    public var separator: Bool?
    public var height: BlockElementHeight?
    public var requires: [String: String]?

    public init(
        id: String? = nil,
        title: String? = nil,
        data: [ChartDataPoint],
        colors: [String]? = nil,
        size: String? = nil,
        showLegend: Bool? = nil,
        showPercentages: Bool? = nil,
        isVisible: Bool? = nil,
        spacing: Spacing? = nil,
        separator: Bool? = nil,
        height: BlockElementHeight? = nil,
        requires: [String: String]? = nil
    ) {
        self.id = id
        self.title = title
        self.data = data
        self.colors = colors
        self.size = size
        self.showLegend = showLegend
        self.showPercentages = showPercentages
        self.isVisible = isVisible
        self.spacing = spacing
        self.separator = separator
        self.height = height
        self.requires = requires
    }
}

File: android/ac-core/src/main/kotlin/com/microsoft/adaptivecards/core/models/ChartElements.kt
package com.microsoft.adaptivecards.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.json.JsonElement

@Serializable
data class ChartDataPoint(
    val label: String,
    val value: Double,
    val color: String? = null
)

@Serializable
@SerialName("DonutChart")
data class DonutChart(
    @Transient override val type: String = "DonutChart",
    override val id: String? = null,
    override val isVisible: Boolean = true,
    override val separator: Boolean = false,
    override val spacing: Spacing? = null,
    override val height: BlockElementHeight? = null,
    override val requires: Map<String, String>? = null,
    override val fallback: JsonElement? = null,
    val title: String? = null,
    val data: List<ChartDataPoint>,
    val colors: List<String>? = null,
    val size: String? = null,
    val showLegend: Boolean? = null,
    val innerRadiusRatio: Double? = null
) : CardElement

@Serializable
@SerialName("BarChart")
data class BarChart(
    @Transient override val type: String = "BarChart",
    override val id: String? = null,
    override val isVisible: Boolean = true,
    override val separator: Boolean = false,
    override val spacing: Spacing? = null,
    override val height: BlockElementHeight? = null,
    override val requires: Map<String, String>? = null,
    override val fallback: JsonElement? = null,
    val title: String? = null,
    val data: List<ChartDataPoint>,
    val colors: List<String>? = null,
    val size: String? = null,
    val showLegend: Boolean? = null,
    val orientation: String? = null,
    val showValues: Boolean? = null
) : CardElement

@Serializable
@SerialName("LineChart")
data class LineChart(
    @Transient override val type: String = "LineChart",
    override val id: String? = null,
    override val isVisible: Boolean = true,
    override val separator: Boolean = false,
    override val spacing: Spacing? = null,
    override val height: BlockElementHeight? = null,
    override val requires: Map<String, String>? = null,
    override val fallback: JsonElement? = null,
    val title: String? = null,
    val data: List<ChartDataPoint>,
    val colors: List<String>? = null,
    val size: String? = null,
    val showLegend: Boolean? = null,
    val showDataPoints: Boolean? = null,
    val smooth: Boolean? = null
) : CardElement

@Serializable
@SerialName("PieChart")
data class PieChart(
    @Transient override val type: String = "PieChart",
    override val id: String? = null,
    override val isVisible: Boolean = true,
    override val separator: Boolean = false,
    override val spacing: Spacing? = null,
    override val height: BlockElementHeight? = null,
    override val requires: Map<String, String>? = null,
    override val fallback: JsonElement? = null,
    val title: String? = null,
    val data: List<ChartDataPoint>,
    val colors: List<String>? = null,
    val size: String? = null,
    val showLegend: Boolean? = null,
    val showPercentages: Boolean? = null
) : CardElement
